### ISSUE TITLE
Is there a way to define a unique constraint on two columns of a table?

### ISSUE THREAD

**Issue by chi6rag** [SimonVT/schematic#54]

**Comment by SimonVT** [comment]
Does adding two `@Unique` annotations not work? What needs to change in the generated SQL?

**Comment by chi6rag** [comment]
I added two uniques like so

```
public class QuoteHistory {
    @DataType(DataType.Type.INTEGER)
    @PrimaryKey
    @AutoIncrement
    public static final String _ID = "_id";
    @DataType(DataType.Type.TEXT)
    @References(table = QuoteDatabase.QUOTES, column = QuoteColumns.SYMBOL)
    @Unique
    public static final String SYMBOL = "symbol";
    @DataType(DataType.Type.TEXT)
    @Unique
    @NotNull
    public static final String DATE = "date";
    @DataType(DataType.Type.TEXT)
    @NotNull
    public static final String OPEN = "open";
    @DataType(DataType.Type.TEXT)
    @NotNull
    public static final String HIGH = "high";
    @DataType(DataType.Type.TEXT)
    @NotNull
    public static final String LOW = "low";
    @DataType(DataType.Type.TEXT)
    @NotNull
    public static final String CLOSE = "close";
    @DataType(DataType.Type.TEXT)
    @NotNull
    public static final String VOLUME = "volume";
    @DataType(DataType.Type.TEXT)
    @NonNull
    public static final String ADJ_CLOSE = "adj_close";
}
```

and the above mentioned DSL generates this:

```
"CREATE TABLE history ("
   + QuoteHistory._ID + " INTEGER PRIMARY KEY AUTOINCREMENT,"
   + QuoteHistory.SYMBOL + " TEXT UNIQUE REFERENCES quotes(symbol),"
   + QuoteHistory.DATE + " TEXT NOT NULL UNIQUE,"
   + QuoteHistory.OPEN + " TEXT NOT NULL,"
   + QuoteHistory.HIGH + " TEXT NOT NULL,"
   + QuoteHistory.LOW + " TEXT NOT NULL,"
   + QuoteHistory.CLOSE + " TEXT NOT NULL,"
   + QuoteHistory.VOLUME + " TEXT NOT NULL,"
   + QuoteHistory.ADJ_CLOSE + " TEXT)"
```

I need to create a `history` table with a unique constraint on a combination of Symbol and Date, so that
If I insert: 
`FB` as symbol and `20052016` as date, it gets successfully inserted
`FB` as symbol and `20052016` as date again, it does not get inserted
`FB` as symbol and `21052016` as date, it gets successfully inserted

**Comment by Vinlaxywei** [comment]
i got same problem like this,did you fix the? @chi6rag

**Comment by SimonVT** [comment]
One way of implementing it could be to have a `@UniqueConstraint` annotation with a name value. Multiple columns could be annotated and the SQL created: https://www.sqlite.org/syntax/table-constraint.html

The annotation should have a conflict-clause value as well, possible forcing the user to specifying the same value for all annotations with the same name.

**Comment by Vinlaxywei** [comment]
Thank for response。

Now i have one columns like this
```
public class TrendColumns {
    @DataType(DataType.Type.INTEGER) @PrimaryKey
    @AutoIncrement
    public static final String _ID = "_id";
    @DataType(DataType.Type.TEXT) @NotNull
    public static final String SYMBOL = "symbol";
    @DataType(DataType.Type.TEXT) @NotNull
    public static final String DATE = "date";
    @DataType(DataType.Type.TEXT) @NotNull
    public static final String BIDPRICE = "bid_price";
}
```

in the app/build/source/QuoteDatabase.java class
```
public static final String TREND = "CREATE TABLE trend ("
            + TrendColumns._ID + " INTEGER PRIMARY KEY AUTOINCREMENT,"
            + TrendColumns.SYMBOL + " TEXT NOT NULL,"
            + TrendColumns.DATE + " TEXT NOT NULL,"
            + TrendColumns.BIDPRICE + " TEXT NOT NULL)";
```

 but i need expected generated table schema like this:
```
" UNIQUE ( date, symbol ) ON CONFLICT REPLACE )";
```

so, if more than one need this function, you may consider to append the

**Comment by SimonVT** [comment]
It'll be a while before I'll have a chance to look at it, but patches welcome :)

**Comment by JiaminZhu** [comment]
Wow I am looking for the exact feature and am happen to see it's being discussed. It's a quite common use case to have multiple columns combined as a unique constraint.

>One way of implementing it could be to have a @UniqueConstraint annotation with a name value. Multiple columns could be annotated and the SQL created: https://www.sqlite.org/syntax/table-constraint.html

>The annotation should have a conflict-clause value as well, possible forcing the user to specifying the same value for all annotations with the same name.

Actually I think you already support multiple primary keys and support 'onConflict=ConflictResolutionType.xxx' as [I searched here](https://github.com/SimonVT/schematic/search?utf8=%E2%9C%93&q=ConflictResolutionType). Do you enforce the conflict resolution to be same for all primary keys? 

By the way, I also see you have @Check annotation in your sample, how to make use of that?  this is simple add arbitrary constrains?  Feeling quite helpless without some help documents.

**Comment by MyDogTom** [comment]
@JiaminZhu 

> Do you enforce the conflict resolution to be same for all primary keys or what was your strategy there?

Please take a look at JavaDoc for [PrimaryKey](https://github.com/SimonVT/schematic/blob/master/schematic-annotations/src/main/java/net/simonvt/schematic/annotation/PrimaryKey.java#L27) annotation.

> Defines conflict resolution algorithm. Will be ignored if there are multiple PrimaryKey annotations

Regarding check constraint.

>Never mind, I just realized that check annotation also only works for one column.

Not true :) You can apply it on table level and provide your custom check constraint with multiple columns. [JavaDoc for Check](https://github.com/SimonVT/schematic/blob/master/schematic-annotations/src/main/java/net/simonvt/schematic/annotation/Check.java#L14) has such example.
```java
@Check(NotesColumns.STATUS + " = '" + NotesColumns.STATUS_COMPLETED + "' and " + NotesColumns.COMPLETION_DATE + " is not null")
   public interface NotesColumns{
     String STATUS_NEW = "new";
     String STATUS_COMPLETED = "completed";
     @DataType(TEXT) @CHECK(NotesColumns.STATUS + " in ('"+NotesColumns.STATUS_NEW+"', '"+NotesColumns.STATUS_COMPLETED+"')") String STATUS = "status";
     @DataType(TEXT) String COMPLETION_DATE = "completion_date";
   }
```

**Comment by JiaminZhu** [comment]
Wow Thank you.  After I tested with several examples, I also noticed that the onConflict resolution algorithm only works with single primary key case. 

So is @Check the only supported table constraint now? unique, primary key and foreign key now only works on column level.  Thanks for you PR that added unique as a table constraint.  looking forward to see it merged!

**Comment by SimonVT** [comment]
Closed by #63
